source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,1896 @@
+nested structures (e.g. individual `test`_, `operation`_).
+
+Although test runners are free to process YAML or JSON files, YAML is the
+canonical format for writing tests. YAML files may be converted to JSON using a
+tool such as `js-yaml <https://github.com/nodeca/js-yaml>`__ .
+
+
+Top-level Fields
+~~~~~~~~~~~~~~~~
+
+The top-level fields of a test file are as follows:
+
+.. _schemaVersion:
+
+- ``schemaVersion``: Required string. Version of this specification to which the
+  test file complies. Test runners will use this to determine compatibility
+  (i.e. whether and how the test file will be interpreted). The format of this
+  string is defined in `Version String`_.
+
+.. _runOn:
+
+- ``runOn``: Optional array of documents. List of server version and/or topology
+  requirements for which the tests in this file can be run. These requirements
+  may be overridden on a per-test basis by `test.runOn <test_runOn_>`_. Test
+  runners MUST skip a test if its requirements are not met.
+
+  If set, the array should contain at least one document. The structure of each
+  document is defined in `runOnRequirement`_.
+
+  **TODO**: Consider whether these requirements should allow an entire test file
+  to be skipped up front, instead of allowing per-test requirements to override.
+
+.. _allowMultipleMongoses:
+
+- ``allowMultipleMongoses``: Optional boolean. If false, all MongoClients
+  created for this test file (internal and any `entities <entity_client_>`_)
+  that could connect to a sharded cluster MUST be initialized with only a single
+  mongos host. Defaults to true. If true or the topology is non-sharded, this
+  option has no effect. Test files that include tests with a `failPoint`_
+  operation that may run on sharded topologies MUST specify false for this
+  option.
+
+.. _createEntities:
+
+- ``createEntities``: Optional array of documents. List of entities (e.g.
+  client, collection, session objects) that should be created before each test
+  case is executed. The structure of each document is defined in `entity`_.
+
+.. _collectionName:
+
+- ``collectionName``: Optional string. Name of collection under test. This is
+  primarily useful when the collection name must be referenced in an assertion.
+  The YAML file SHOULD define a `node anchor`_ for this field (e.g.
+  ``collectionName: &collectionName foo``). If unset, test runners may use
+  whatever value they prefer.
+
+.. _databaseName:
+
+- ``databaseName``: Optional string. Name of database under test. This is
+  primarily useful when the database name must be referenced in an assertion.
+  The YAML file SHOULD define a `node anchor`_ for this field (e.g.
+  ``databaseName: &databaseName foo``). If unset, test runners may use whatever
+  value they prefer.
+
+.. _initialData:
+
+- ``initialData``: Optional array of documents. Data that should exist in
+  collections before each test case is executed.
+
+  If set, the array should contain at least one document. The structure of each
```

Comment:
I think your question is about `documents` in `collectionData`, since that's what might actually be passed to `insert_many`. This text is just saying that `initialData` should at least contain one element (i.e. some instructions for a particular collection).

Quoting the current text of "Executing a Test":

> If `initialData`_ is specified, for each `collectionData`_ therein the test runner MUST drop the collection using a "majority" write concern and then insert the specified documents, also using a "majority" write concern. The test runner MUST use the internal MongoClient for these operations.

Given the way that's worded, if `initialData` is empty there would be nothing to do. I also used "should" in this way for other fields where it makes no sense to leave them empty (e.g. `runOn`). I do think "SHOULD" or "MUST" would be more appropriate. FWIW, I plan on enforcing a minimum of 1 element via the JSON schema.

On a related note, and more to your original question about what gets passed to `insert_many`, I _do_ think it's a good idea to allow `documents` to be an empty array. In that case, the collection should still get dropped and I'd amend the language above to clarify that no documents should be inserted (i.e. skip `insert_many`).

Assuming collections are only dropped from processing of `initialData` (I don't instruct test runners to drop collections elsewhere), this would address the need in [SPEC-1723](https://jira.mongodb.org/browse/SPEC-1723) to disable drop behavior without needing an extra option on a collection entity definition (akin to `createOnServer`).

Reply:
Will clarify that test runners MUST not call `insert_many` if `documents` is empty. If empty, test runners MUST explicitly create the collection with a majority write concern, and then we can do away with `createOnServer`.

Reply:
Text in "Executing a Test" is now updated. I removed `createOnServer` and rewrote the section on processing `initialData`. If no documents are specified, I instruct test runners to just create the collection (instead of inserting).